src/pgcopy/src/copy.rs

```diff
@@ -93,6 +105,62 @@ pub fn encode_copy_row_text(
     Ok(())
 }
 
+pub fn encode_copy_row_csv(
+    row: Row,
+    typ: &RelationType,
+    out: &mut Vec<u8>,
+    CopyCsvFormatParams {
+        delimiter: delim,
+        quote,
+        escape,
+        header: _,
+        null,
+    }: CopyCsvFormatParams,
+) -> Result<(), io::Error> {
+    let null = null.as_bytes();
+    let is_special = |c: &u8| *c == delim || *c == quote || *c == b'\r' || *c == b'\n';
+    let mut buf = BytesMut::new();
+    for (idx, field) in mz_pgrepr::values_from_row(row, typ).into_iter().enumerate() {
+        if idx > 0 {
+            out.push(delim);
+        }
+        match field {
+            None => out.extend(null),
+            Some(field) => {
+                buf.clear();
+                field.encode_text(&mut buf);
+                // A field needs quoting if:
+                //   * It is the only field and the value is exactly the end
+                //     of copy marker.
+                //   * The field contains a special character.
+                //   * The field is exactly the NULL sentinel.
```

Comment:
I'm not sure of this. https://www.postgresql.org/docs/current/sql-copy.html says

> If the value contains the delimiter character, the QUOTE character, the NULL string, a carriage return, or line feed character, then the whole value is prefixed and suffixed by the QUOTE character...

Reply:
FYI this code is based on a WIP branch I have from a while back where I tried to fix up the `pgcopy` crate to handle all three formats in both directions. Never got around to finishing it but the CSV encoding path was complete: https://github.com/MaterializeInc/materialize/compare/v0.68.3...benesch:materialize:copy-more?expand=1#diff-7a9592753ee463a8c8b47910f19f2ed6dd0c76d533831dd86b629b255d61a376R101

Which part are you not sure of? The end of copy marker special case is mentioned a few paragraphs done from the one you quoted:

> Because backslash is not a special character in the CSV format, \., the end-of-data marker, could also appear as a data value. To avoid any misinterpretation, a \. data value appearing as a lone entry on a line is automatically quoted on output, and on input, if quoted, is not interpreted as the end-of-data marker. If you are loading a file created by another application that has a single unquoted column and might have a value of \., you might need to quote that value in the input file.

Reply:
The comment here says "is exactly the NULL sentinel" and `is_special` doesn't consult that. The spec says "contains ... the NULL string". The difference to my read is between "is exactly" and "contains".

Reply:
My memory is hazy, but I seem to recall that the spec is wrong on this point, and PostgreSQL only quotes if the value exactly matches the NULL string. AFK so can’t test.